spec.bs

```diff
@@ -232,7 +232,161 @@ to conduct an auction to select an advertisement to display to the user, they ca
+1. If |generateBidOutput|["{{GenerateBidOutput/ad}}"] [=map/exists=]:
+  1. Let |adJson| be the result of [=serializing a JavaScript value to a JSON string=], given
+    |generateBidOutput|["{{GenerateBidOutput/ad}}"].
+  1. If |adJson| is failure, return failure.
+  1. Set |bidToSet|'s [=generated bid/ad=] to |adJson|.
+1. Let |renderURL| be null.
+1. If |generateBidOutput|["{{GenerateBidOutput/adDescriptor}}"] is a {{DOMString}}:
+  1. Let |renderURL| be the result of running the [=URL parser=] on
+    |generateBidOutput|["{{GenerateBidOutput/adDescriptor}}"].
+1. Otherwise:
+  1. TODO: Let |adDescriptor| be the result of [=converting an ad descriptor=] given
+    |generateBidOutput|["{{GenerateBidOutput/adDescriptor}}"].
+  1. If |adDescriptor| is failure, return false.
+  1. Let |parsedAdDescriptor| be the result of parsing |adDescriptor|.
+  1. If |parsedAdDescriptor| is failure, return false.
+  1. Let |renderURL| be the result of running the [=URL parser=] on |parsedAdDescriptor|'s render.
+1. If |renderURL| is not an allowed ad url (TODO: expand), return false.
+1. Let |adComponents| be |generateBidOutput|["{{GenerateBidOutput/adComponentDescriptors}}"].
+1. If |adComponents| is not null or undefined:
+  1. If biddingScriptRunner's adComonents parameter does not exist, return false.
+1. If |adComponents| is not an array, return false.
+1. If |adComponents|'s size is greater than MaxAdAuctionAdComponents (20, TODO: define this as a
+  vendor specific value?), return false.
+1. Let |internalAdComponents| be a new [=list=] of [=ad descriptors=].
+1. For |component| in |adComponents|:
+  1. Let |internalComponent| be a new [=ad descriptor=].
+  1. If |component| is string:
+    1. Set |internalComponent|'s [=ad descriptor/url=] to |component|.
+    1. Set |internalComponent|'s [=ad descriptor/size=] to null.
+  1. Otherwise:
+    1. Let |parsedComponent| be the result of [=converting an ad descriptor=] given |component|.
+    1. If |parsedComponent|'s adSize is invalid, return false.
+    1. If |parsedComponent|'s render is not allowed, return false.
+    1. Set |internalComponent| to |parsedComponent|.
+  1. [=list/Append] |internalComponent| to |internalAdComponents|.
+1. Set |bidToSet|'s [=generated bid/ad component descriptors=] to |internalAdComponents|.
+1. If |generateBidOutput|["{{GenerateBidOutput/adCost}}"] [=map/exists=]:
+  1. Set |bidToSet|'s [=generated bid/ad cost=] to
+    |generateBidOutput|["{{GenerateBidOutput/adCost}}"].
+1. If |generateBidOutput|["{{GenerateBidOutput/modelingSignals}}"] [=map/exists=]:
+  1. Let |modelingSignals| be |generateBidOutput|["{{GenerateBidOutput/modelingSignals}}"].
+  1. If |modelingSignals| is greater than 0 and less than 4096:
+    1. Set |bidToSet|'s [=generated bid/modeling signals=] to |modelingSignals|
+1. Set bid to |bidToSet|.
+1. Return true.
+
+To <dfn>convert an ad descriptor</dfn> given an {{AdDescriptor}} |adDescriptor|:

```

Comment:
Should this start with "Let |parsedAdDescriptor| be a new [=ad descriptor=]" and end with "return |parsedAdDescriptor|"?